# TITLE
Coupling microbiome to functions diversity in waste water treatment ecosystems
  
# SUMMARY
The main aim of the research is to carry an extensive microbiome analysis using a metaomics dataset according to conventional microbiome analysis provided in Bioconductor in R (Refer to guiding online book: orchestrating microbiome analysis). The dataset is made up of metagenomics and meta-transcriptomic data. The data is a result of a longitudinal sampling campaign in the aeration tank in a saline wastewater treatment plant which is operated using activated sludge microbiome. RNA extractions were taken from 32 different samples which were collected from 2014 until 2017. The samples were taken 2 to 3 weeks apart. The data represents an opportunity to connect microbiome and function diversity patterns by connecting OTUs, metagenomics and meta-transcriptomic data across time.

# QUESTIONS (To be refined)
  * Do abiotic factors affect the interplay between microbiome and function diversity? 
  * Does the coupling between microbiome and function diversity flucuate in time?
  * Are there invariants in the microbiome-function diversity coupling? 

# DATA
https://wetsus-my.sharepoint.com/personal/amah_wetsus_nl/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Famah%5Fwetsus%5Fnl%2FDocuments%2FOosterhorn%5FDataDescriptor&ga=1

# TODO
  * Functions (multiple regression, ML) underlying abiotic factors predicting microbiome, metagenomics and meta-transcriptomic data
  * Coupling microbiome and metabolic pathways diversity
  * Jupyter notebook including R analysis
  * Diversity-function matrix (OTUs composition and prevalance vs. metabolic pathways): Does it significantly differ from idiosyncratic (randomly) assembly?
